### PR TITLE
hardhat + ethers v5

### PR DESCRIPTION
**Pull request by acrylix**
changes:
- hardhat
- ethers v5
- solc 0.7.6
- enable yul optimizer in solc settings

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "wallet-contracts",
-  "version": "1.0.1",
+  "version": "2.0.0",
```
Comment (acrylix) [inline package.json:3]: bumping major version here due to slight changes in bytecode from adjusted solc optimizer settings, please amend if this should be a minor version update instead

File: hardhat.config.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import 'solidity-coverage'
+
+const ganacheNetwork = {
+  url: 'http://127.0.0.1:8545',
+  blockGasLimit: 6000000000
+}
+
+const config: HardhatUserConfig = {
+  solidity: {
+    version: '0.7.4',
```
Comment (pkieltyka) [inline hardhat.config.ts:17]: lets use 0.7.6

File: package.json [hunk 3]
```diff
@@ -1,23 +1,22 @@
...
+    "test": "hardhat test",
+    "quicktest": "hardhat test",
     "precoverage": "yarn build",
-    "coverage": "COVERAGE=true NET_ID=1 buidler coverage --network coverage",
+    "coverage": "COVERAGE=true NET_ID=1 hardhat coverage --network coverage",
     "benchmark": "BENCHMARK=true yarn test",
     "gen:contract-typings": "rimraf typings/contracts/* && yarn gen:contract-typings:ethers4 && yarn gen:contract-typings:ethers5",
-    "gen:contract-typings:ethers4": "typechain --target ethers-v4 --outDir typings/contracts/ethers-v4 './artifacts/*.json'",
-    "gen:contract-typings:ethers5": "typechain --target ethers-v5 --outDir typings/contracts/ethers-v5 './artifacts/*.json'",
+    "gen:contract-typings:ethers4": "typechain --target ethers-v4 --outDir typings/contracts/ethers-v4 './artifacts/!(build-info)/**/*[^dbg].json'",
```
Comment (pkieltyka) [inline package.json:18]: any reason to keep supporting ethers-v4 typings?

File: package.json [hunk 4]
```diff
@@ -28,15 +27,15 @@
...
+    "deploy:ganache": "hardhat run --network ganache src/scripts/deploy-contracts.ts",
+    "deploy": "hardhat run src/scripts/deploy-contracts.ts --network"
   },
   "files": [
     "artifacts",
     "typings"
   ],
   "config": {
-    "mnemonic": "concert load couple harbor equip island argue ramp clarify fence smart topic",
+    "mnemonic": "test test test test test test test test test test test junk",
```
Comment (pkieltyka) [inline package.json:38]: why? shouldn't we either remove this "config" entry, and use hardhat.config.ts mnemonic, or put something that works?

**Review by pkieltyka — COMMENTED** [review]
nice quick work on this PR!

File: package.json (hunk 1 above)
Comment (pkieltyka) [reply package.json:3]: @acrylix I think lets reserve v2 for later, how just v1.1.0 for now?

**Review by pkieltyka — COMMENTED** [review]
In addition, I think we should remove "artifacts" from .gitignore, so that we actually commit the artifacts to git

**Comment by pkieltyka** [comment]
as well lets add "networks" and "src" to the "files" array of package.json

File: package.json (hunk 3 above)
Comment (acrylix) [reply package.json:18]: I think this was kept for backward compatibility, if no objections lets remove v4 typings and only support v5

File: package.json (hunk 4 above)
Comment (acrylix) [reply package.json:38]: this is for ganache

Comment (pkieltyka) [reply package.json:39]: Does that “test test .. “ work on ganache?

Comment (acrylix) [reply package.json:39]: yea its valid mnemonic, its nice because it's self explanatory that it is a test config

**Comment by acrylix** [comment]
> In addition, I think we should remove "artifacts" from .gitignore, so that we actually commit the artifacts to git

any specific reason that we should? similar to typechain typings, we should have them compiled/generated post-install
https://github.com/ethereum-ts/TypeChain#q-should-i-commit-generated-files-to-my-repository

as long as the solc version and settings are the same, the compile output should be deterministic

File: package.json (hunk 4 above)
Comment (pkieltyka) [reply package.json:39]: Neat

**Comment by pkieltyka** [comment]
I think typings and artifacts should be included in version control cuz output is deterministic and we can keep things faster when consuming the library. I don’t see a downside from including it, and also safer to know source of truth with history to a tag etc

**Comment by pkieltyka** [comment]
I believe the opinion from ethereum-ts and typechain if we should version control them is just wrong. It's better to have them version controlled for reference and use, and if someone prefers they can regen. Same reason we don't generate webrpc gen files on-demand for example. Code generated code should be version controlled even if its reproducible.

I know it's holiday time so no rush haha, but just wanted to leave those comments. I will update it tomm morning. Thanks for this awesome PR

**Comment by acrylix** [comment]
> I believe the opinion from ethereum-ts and typechain in the link above is just wrong. It's better to have them version controlled for reference, dist and use, and if someone prefers they can regen. Same reason we don't generate webrpc gen files on-demand for example. Code generated code should be version controlled even if its reproducible.
> 
> I know it's holiday time so no rush haha, but just wanted to leave those comments. I will update it tomm morning. Thanks for this awesome PR

makes sense, we should make this change in other contract repos as well, I’ll find some time to do that next

**Comment by pkieltyka** [comment]
FYI, I pushed a comment with a few changes:
* updated .gitignore to keep artifacts and typings
* updated package.json slightly, see file
* included artifacts + typings to version control as discussed
* renamed pkg to @0xsequence/wallet-contracts

**Comment by acrylix** [comment]
@pkieltyka LGTM

**Pull request merged**